oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/resource/web/DefaultBearerTokenResolver.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.resource.web;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.springframework.security.oauth2.resource.BearerTokenAuthenticationException;
+import org.springframework.security.oauth2.resource.BearerTokenError;
+import org.springframework.security.oauth2.resource.BearerTokenErrorCodes;
+import org.springframework.util.StringUtils;
+
+/**
+ * The default {@link BearerTokenResolver} implementation based on RFC 6750.
+ *
+ * @author Vedran Pavic
+ * @since 5.1
+ * @see <a href="https://tools.ietf.org/html/rfc6750#section-2" target="_blank">RFC 6750 Section 2: Authenticated Requests</a>
+ */
+public final class DefaultBearerTokenResolver implements BearerTokenResolver {
```

Comment:
I would prefer that we have a separate implementation for each resolving strategy. For example, the `DefaultBearerTokenResolver` becomes `AuthorizationHeaderBearerTokenResolver` and this is the only implementation we provide for now.

Later on we can add `RequestBodyBearerTokenResolver` and `QueryParameterBearerTokenResolver` implementation if need be. NOTE: Both of these are `optional` as per spec and I doubt we'll provide a `QueryParameterBearerTokenResolver` implementation but the user will have the ability to provide their own and configure it.

Reply:
I don't believe that would be a good idea since we would be providing `BearerTokenResolver` implementations that violate the spec. The [RFC 6749 Section 2.1](https://tools.ietf.org/html/rfc6750#section-2.1) clearly states that Authorization header method:

> Clients SHOULD make authenticated requests with a bearer token using
   the "Authorization" request header field with the "Bearer" HTTP
   authorization scheme.  Resource servers MUST support this method.

Other methods can be supported on top, not instead of Authorization header method so the proposed implementation of `DefaultBearerTokenResolver` closely aligns with the spec.

Reply:
I think we are talking the same thing. To re-clarify, we provide one implementation of `BearerTokenResolver` only, this being the Authorization Request Header as it's required per spec. However, instead of naming it `DefaultBearerTokenResolver`, we rename it to be more specific to the strategy `AuthorizationHeaderBearerTokenResolver`.

Reply:
I'm not following. In the previous comment, you mentioned adding `RequestBodyBearerTokenResolver` and `QueryParameterBearerTokenResolver` later on - those two clearly don't support the Authorization header method which is required by the spec.

I'd again like to stress that other token resolution methods are be supported on top of required Authorization header method, rather than replacing it as that would render resource server not compliant with the spec.

I also think we really should offer optional support of other methods as the effort is minimal, the old `spring-security-oauth2` also has that, and if you're building a resource server that's consumed by a variety of clients (i.e. not all are under your control) it's likely there will be older clients which use other methods.

Reply:
@jgrandja, if we rename it to `AuthorizationHeaderBearerTokenResolver`, are you also saying to remove the support already in the class for the other two strategies? The name seems a bit odd if we are leaving that functionality in.

And then, are you also suggesting we not provide support (initially) for the other two strategies?

> Other methods can be supported on top, not instead of Authorization header method

@vpavic, if I'm not mistaken, @jgrandja  is not suggesting that `RequestBodyBearerTokenResolver` would only resolve via the request body, but that it would first delegate to `AuthoriationHeaderBearerTokenResolver`, still then complying with the spec. I believe he is simply stating that the configuration you are going for can be achieved via a specialized class instead of via a flag.

For me, there is value in both of these approaches, the `Default` approach concisely addresses the spec all at once. The `AuthorizationHeader` approach is impossible to accidentally misconfigure--i.e. to get `RequestBody` support, a dev would need to change out the implementation, not just get a configuration flag incorrect.

Reply:
> is not suggesting that `RequestBodyBearerTokenResolver` would only resolve via the request body, but that it would first delegate to `AuthoriationHeaderBearerTokenResolver`

It's not clear to me how would you make things work with this approach if you want to enable both  optional methods (form post body param & query param)? Also keep in mind that the spec states that _Clients MUST NOT use more than one method to transmit the token in each request_ and that `invalid_request` should be returned if they do. I don't see an easy way how to meet this requirement with different `BearerTokenResolver` without introducing significant amount of complexity.

> The `AuthorizationHeader` approach is impossible to accidentally misconfigure--i.e. to get `RequestBody` support, a dev would need to change out the implementation, not just get a configuration flag incorrect.

If currently proposed `DefaultBearerTokenResolver` is registered by default, to enable optionally supported methods one would need explicitly configure and register their own `DefaultBearerTokenResolver` instance - IMO that's pretty robust as well and shouldn't be easy to misconfigure by accident.

Reply:
> If currently proposed `DefaultBearerTokenResolver` is registered by default, to enable optionally supported methods one would need explicitly configure and register their own `DefaultBearerTokenResolver` instance.

I'll defer to @jgrandja on this point: Is default config sufficient coverage against accidental misconfiguration?

Reply:
@jzheaux  
> if we rename it to AuthorizationHeaderBearerTokenResolver, are you also saying to remove the support already in the class for the other two strategies? The name seems a bit odd if we are leaving that functionality in.

Correct, we remove the support for the other two strategies for `AuthorizationHeaderBearerTokenResolver`.

> And then, are you also suggesting we not provide support (initially) for the other two strategies?

I feel that we should focus our initial effort on `AuthorizationHeaderBearerTokenResolver` (the required one) and log a separate ticket for Request Body strategy and deal with that later. The Uri query parameter strategy has plenty of security deficiencies that I don't feel we should provide an implementation for it.

@vpavic 

> It's not clear to me how would you make things work with this approach if you want to enable both optional methods (form post body param & query param)...

I obviously did not do a good job communicating my points across so I figure code snippets will be the most clear. This is what I'm thinking...


```
public interface BearerTokenResolver {
}

// Required implementation
public class AuthorizationHeaderBearerTokenResolver implements BearerTokenResolver {
}

// Optional implementation
public class RequestBodyBearerTokenResolver implements BearerTokenResolver {
}

// Optional implementation NOT RECOMMENED
public class QueryParameterBearerTokenResolver implements BearerTokenResolver {
}

public class DelegatingBearerTokenResolver implements BearerTokenResolver {
	// Required as per spec
	private BearerTokenResolver defaultBearerTokenResolver = new AuthorizationHeaderBearerTokenResolver();

	// Optional as per spec
	private Set<BearerTokenResolver> bearerTokenResolvers;

	@Override
	public String resolve(HttpServletRequest request) {
		// Delegate to bearerTokenResolvers (if available)

		// If bearer token IS NOT resolved from (optional) bearerTokenResolvers than delegate to defaultBearerTokenResolver

		// If bearer token IS resolved from (optional) bearerTokenResolvers than also delegate to defaultBearerTokenResolver
		// to see if it resolves there as well. For example, if it resolves more than once than throw exception as per spec
     
	}
}
```

As far as configuration goes, let's assume a user would configure a custom `BearerTokenResolver` via `httpSecurity.oauth2().resourceServer().tokenResolver()`.

The `SecurityConfigurer` will always create a `DelegatingBearerTokenResolver` and configure it with the one supplied by the user as an additional resolver. The main point here is that the `SecurityConfigurer` will always create a `DelegatingBearerTokenResolver` and therefore the default will always be `AuthorizationHeaderBearerTokenResolver`.

The delegating strategy is a common design pattern that is already used throughout the framework and provides great flexibility and adaptability in it's design.

Reply:
Thanks for the clarification @jgrandja. The delegating resolver approach would've been valid if the bearer token transport methods were fully exclusive.

I'll emphasize some aspects of RFC 6750 from my previous comment one more time:

- https://tools.ietf.org/html/rfc6750#section-2:
> This section defines three methods of sending bearer access tokens in resource requests to resource servers. **Clients MUST NOT use more than one method to transmit the token in each request.**

- https://tools.ietf.org/html/rfc6750#section-3.1:
> invalid_request
The request is missing a required parameter, includes an unsupported parameter or parameter value, repeats the same parameter, **uses more than one method for including an access token**, or is otherwise malformed.  The resource server SHOULD respond with the HTTP 400 (Bad Request) status code.

The pseudo-code presented above not only violates this by completely ignoring presence of optional methods by default and thus not detecting use of multiple methods, but it also actually prefers the optional method(s) to the required one. To satisfy the requirements quote above, the server needs to parse all 3 methods anyway and at that point it's only about what to do with them.

Reply:
@vpavic 

> The pseudo-code presented above...

This is strictly pseudo-code and has not been fully thought out so please don't take it as a literal implementation. My purpose was to flush out the design.

Again, as mentioned, the design presented is a common pattern that is being used throughout the framework and is quite flexible as far as plugging in other implementations and gives the "delegating" construct the ability of a "controller". 

> The delegating resolver approach would've been valid if the bearer token transport methods were fully exclusive.

The implementation of `DelegatingBearerTokenResolver` will work and will fulfil the spec requirements. Sounds like you don't agree?

> To satisfy the requirements quoted above, the server needs to parse all 3 methods anyway and at that point it's only about what to do with them.

Why can't the `DelegatingBearerTokenResolver` parse all internal resolvers to determine if more than one method is used by the client? It acts as a controller and it therefore has this control/ability to do so.

Reply:
> The implementation of `DelegatingBearerTokenResolver` will work and will fulfil the spec requirements. Sounds like you don't agree?

From the approach you're describing I don't believe that's not the case.

A request that presents Bearer Token using multiple transport methods should be treated as invalid regardless of whether Resource Server would actually use methods other than Authorization header. With optional `Set<BearerTokenResolver>` approach you wouldn't be looking at other methods unless appropriate `BearerTokenResolver` is registered.

Reply:
> A request that presents Bearer Token using multiple transport methods should be treated as invalid

Again, the `optional Set<BearerTokenResolver> ` is pseudo-code and should not be taken literally as an implementation. The focus is on the adaptability/flexibility of the design, implementation needs to get flushed out.

If `DelegatingBearerTokenResolver` had both `AuthorizationHeaderBearerTokenResolver` and `RequestBodyBearerTokenResolver` configured than it can fulfill the spec requirement:

> invalid_request
         The request is missing a required parameter, includes an
         unsupported parameter or parameter value, repeats the same
         parameter, **uses more than one method for including an access
         token**, or is otherwise malformed.  The resource server SHOULD
         respond with the HTTP 400 (Bad Request) status code.

I do not want to provide support for URI Query Parameter method based on a number of Security deficiencies outlined in the spec and that are well-known when transporting sensitive information in URI parameters.

Reply:
@vpavic As mentioned, @rwinch agrees (during our standup) that the current implementation of `DefaultBearerTokenResolver` is the way to go over the delegating strategy along with providing support for the 3 authentication methods. Main thing is to disable form and query by default but allow for configuration if the user requires it. So basically let's go with what you have.